overlord/hookstate/ctlcmd/model.go

```diff
@@ -0,0 +1,280 @@
+func (c *modelCommand) reportError(format string, a ...interface{}) {
+	w := c.newTabWriter(c.stderr)
+	fmt.Fprintf(w, format, a...)
+	w.Flush()
+}
+
+// hasSnapdControl returns true if the requesting snap has the snapd-control plug
+// and only if it is connected as well.
+var hasSnapdControlInterface = func(st *state.State, snapName string) (bool, error) {
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return false, err
+	}
+	for refStr, connState := range conns {
+		if connState.Undesired || connState.Interface != "snapd-control" {
+			continue
+		}
+		connRef, err := interfaces.ParseConnRef(refStr)
+		if err != nil {
+			return false, err
+		}
+		if connRef.PlugRef.Snap == snapName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// getSnapInfoWithPublisher is a helper utility to read the snap.Info for the requesting snap
+// which also fills the publisher information.
+func (c *modelCommand) getSnapInfoWithPublisher(st *state.State, snapName string) (*snap.Info, error) {
+	var snapst snapstate.SnapState
+	if err := snapstate.Get(st, snapName, &snapst); err != nil {
+		return nil, fmt.Errorf("failed to get snapstate for snap %s: %v", snapName, err)
+	}
+
+	snapInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	snapInfo.Publisher, err = assertstate.PublisherStoreAccount(st, snapInfo.SnapID)
+	return snapInfo, err
+}
+
+// checkPermissions verifies that the snap described by snapInfo is allowed to
+// read the model assertion of deviceCtx.
+// We allow the usage of this command if one of the following is true
+// 1. The requesting snap must be a gadget
+// 2. Come from the same brand as the device model assertion
+// 3. Have the snapd-control plug
+func (c *modelCommand) checkPermissions(st *state.State, deviceCtx snapstate.DeviceContext, snapInfo *snap.Info) error {
+	if snapType := snapInfo.Type(); snapType == snap.TypeGadget {
+		return nil
+	}
+	if snapInfo.Publisher.ID == deviceCtx.Model().BrandID() {
+		return nil
+	}
+	if conn, err := hasSnapdControlInterface(st, snapInfo.SnapName()); err != nil {
+		return fmt.Errorf("cannot check for snapd-control interface: %v", err)
+	} else if conn {
+		return nil
+	}
+
+	c.reportError("cannot get model assertion for snap %q: "+
+		"must be either a gadget snap, from the same publisher as the model "+
+		"or have the snapd-control interface\n", snapInfo.SnapName())
+	return fmt.Errorf("insufficient permissions to get model assertion for snap %q", snapInfo.SnapName())
+}
+
+// findSerialAssertion is a helper function to find the newest matching serial assertion
+// for the provided model assertion.
+func findSerialAssertion(st *state.State, modelAssertion *asserts.Model) (*asserts.Serial, error) {
+	assertions, err := assertstate.DB(st).FindMany(asserts.SerialType, map[string]string{
+		"brand-id": modelAssertion.BrandID(),
+		"model":    modelAssertion.Model(),
+	})
+	if err != nil || len(assertions) == 0 {
+		return nil, err
+	}
+
```

Comment:
the rest of this function seems also not exercised

Reply:
Thanks again @pedronis ! I have added tests for this now